core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1018,7 +1025,9 @@ private[spark] class BlockManager(
           try {
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.dispose()
+            if (!level.useOffHeap) {
```

Comment:
I agree, we can have `doGetLocalBytes` return an additional `boolean` that indicates whether the returned buffer is memory-mapped or not.